pkg/operator/controller/ingress/namespace.go

```diff
@@ -4,26 +4,128 @@ import (
+	operatorcontroller "github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
 )
 
-// ensureRouterNamespace ensures all the necessary scaffolding exists for
-// routers generally, including a namespace and all RBAC setup.
-func (r *reconciler) ensureRouterNamespace() error {
-	ns := manifests.RouterNamespace()
-	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
-		if !errors.IsNotFound(err) {
-			return fmt.Errorf("failed to get router namespace %q: %v", ns.Name, err)
+// ensureRouterNamespace ensures that the router namespace exists.
+func (r *reconciler) ensureRouterNamespace() (bool, *corev1.Namespace, error) {
+	desired := manifests.RouterNamespace()
+
+	haveNamespace, current, err := r.currentRouterNamespace()
+	if err != nil {
+		return false, nil, err
+	}
+
+	switch {
+	case !haveNamespace:
+		if err := r.client.Create(context.TODO(), desired); err != nil {
+			return false, nil, fmt.Errorf("failed to create router namespace: %v", err)
+		}
+		log.Info("created router namespace", "desired", desired)
+		return r.currentRouterNamespace()
+	case haveNamespace:
+		if updated, err := r.updateRouterNamespace(current, desired); err != nil {
+			return true, current, fmt.Errorf("failed to update router namespace: %v", err)
+		} else if updated {
+			return r.currentRouterNamespace()
+		}
+	}
+	return true, current, nil
+}
+
+// currentRouterNamespace gets the current router namespace resource.
+func (r *reconciler) currentRouterNamespace() (bool, *corev1.Namespace, error) {
+	namespace := &corev1.Namespace{}
+	name := types.NamespacedName{
+		Name: operatorcontroller.DefaultOperandNamespace,
+	}
+	if err := r.client.Get(context.TODO(), name, namespace); err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
+	return true, namespace, nil
+}
+
+// updateRouterNamespace updates the router namespace if an appropriate change
+// has been detected.
+func (r *reconciler) updateRouterNamespace(current, desired *corev1.Namespace) (bool, error) {
+	changed, updated := routerNamespaceChanged(current, desired)
+	if !changed {
+		return false, nil
+	}
+	// Diff before updating because the client may mutate the object.
+	diff := cmp.Diff(current, updated, cmpopts.EquateEmpty())
+	if err := r.client.Update(context.TODO(), updated); err != nil {
+		if errors.IsAlreadyExists(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	log.Info("updated router namespace", "namespace", updated.Name, "diff", diff)
+	return true, nil
+}
+
+// routerNamespaceChanged returns true if current and expected differ by any
+// of the annotations or the labels controlled by the operator.
+// Other namespace labels and annotations are not overwritten so that
+// values written by other controllers are preserved.
+func routerNamespaceChanged(current, expected *corev1.Namespace) (bool, *corev1.Namespace) {
+	knownAnnotations := []string{
+		"openshift.io/node-selector",
```

Comment:
This one worries me a little.  For 4.8, would it be worth setting `Upgradeable=False` if the administrator has set this annotation, and then reconcile it only in 4.9+?

Reply:
What if the operator did not manage this annotation at all?

Reply:
Removed the "openshift.io/node-selector" annotation here as per a conversation on slack.

Reply:
Yeah, if it isn't needed to fix [BZ#1954330](https://bugzilla.redhat.com/show_bug.cgi?id=1954330), we can drop reconciliation of that annotation from this PR.